File: test/features/ocr/amount_extractor_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:war2aty/features/ocr/domain/services/amount_extractor.dart';

void main() {
  const extractor = AmountExtractor();

  group('AmountExtractor', () {
    group('explicit currency', () {
      test('number + EGP', () {
        final results = extractor.extract('250.50 EGP');
        expect(results, hasLength(1));
        expect(results.first.value, equals(250.50));
        expect(results.first.currency, equals('EGP'));
        expect(results.first.isAmbiguous, isFalse);
      });

      test('EGP + number', () {
        final results = extractor.extract('EGP 1000');
        expect(results, hasLength(1));
        expect(results.first.value, equals(1000));
        expect(results.first.currency, equals('EGP'));
      });

      test('number + LE', () {
        final results = extractor.extract('125.00 LE');
        expect(results, hasLength(1));
        expect(results.first.currency, equals('EGP'));
      });

      test('number with commas', () {
        final results = extractor.extract('1,250.50 EGP');
        expect(results, hasLength(1));
        expect(results.first.value, equals(1250.50));
      });

      test('dollar sign', () {
        final results = extractor.extract('\$100');
        expect(results, hasLength(1));
        expect(results.first.currency, equals('USD'));
      });

      test('euro sign', () {
        final results = extractor.extract('€50.00');
        expect(results, hasLength(1));
        expect(results.first.currency, equals('EUR'));
      });
    });

    group('keyword-adjacent', () {
      test('إجمالي + number', () {
        final results = extractor.extract('إجمالي 250.50');
        expect(results, hasLength(1));
        expect(results.first.value, equals(250.50));
        expect(results.first.isAmbiguous, isTrue);
      });

      test('المطلوب: number', () {
        final results = extractor.extract('المطلوب: 1500');
        expect(results, hasLength(1));
        expect(results.first.value, equals(1500));
        expect(results.first.isAmbiguous, isTrue);
      });

      test('القيمة number', () {
        final results = extractor.extract('القيمة 300');
        expect(results, hasLength(1));
      });

      test('مبلغ number', () {
        final results = extractor.extract('مبلغ 500');
        expect(results, hasLength(1));
      });

      test('ضريبة number', () {
        final results = extractor.extract('ضريبة 14.00');
        expect(results, hasLength(1));
        expect(results.first.value, equals(14.00));
      });

      test('خصم number', () {
        final results = extractor.extract('خصم 50');
        expect(results, hasLength(1));
      });
    });

    group('dedup', () {
      test('does not duplicate when explicit and keyword overlap', () {
        final results = extractor.extract('إجمالي 250.50 EGP');
        expect(results, hasLength(1));
        expect(results.first.currency, equals('EGP'));
        expect(results.first.isAmbiguous, isFalse);
      });
    });

    group('multiple amounts', () {
      test('extracts multiple amounts in one text', () {
        final results = extractor.extract('ضريبة 14.00 EGP\nإجمالي 250');
        expect(results, hasLength(2));
      });
    });

    group('no false positives', () {
      test('does not extract bare numbers without context', () {
        final results = extractor.extract('رقم الحساب 123456');
        expect(results, isEmpty);
      });
    });
  });
}

File: supabase/tests/unit/amount-extractor.test.ts
/**
 * F13-T05 · Tests for the amount extractor.
 *
 * Mirrors `test/features/ocr/amount_extractor_test.dart` case-for-case.
 */

import { assertEquals } from "jsr:@std/assert@1";

import { extractAmounts } from "../../functions/_shared/extractors/amount-extractor.ts";

// ── explicit currency ────────────────────────────────────────────────────────

Deno.test("amount: number + EGP", () => {
  const results = extractAmounts("250.50 EGP");
  assertEquals(results.length, 1);
  assertEquals(results[0].value, 250.5);
  assertEquals(results[0].currency, "EGP");
  assertEquals(results[0].is_ambiguous, false);
});

Deno.test("amount: EGP + number", () => {
  const results = extractAmounts("EGP 1000");
  assertEquals(results.length, 1);
  assertEquals(results[0].value, 1000);
  assertEquals(results[0].currency, "EGP");
});

Deno.test("amount: number + LE", () => {
  const results = extractAmounts("125.00 LE");
  assertEquals(results.length, 1);
  assertEquals(results[0].currency, "EGP");
});

Deno.test("amount: number with commas", () => {
  const results = extractAmounts("1,250.50 EGP");
  assertEquals(results.length, 1);
  assertEquals(results[0].value, 1250.5);
});

Deno.test("amount: dollar sign", () => {
  const results = extractAmounts("$100");
  assertEquals(results.length, 1);
  assertEquals(results[0].currency, "USD");
});

Deno.test("amount: euro sign", () => {
  const results = extractAmounts("€50.00");
  assertEquals(results.length, 1);
  assertEquals(results[0].currency, "EUR");
});

// ── keyword-adjacent ─────────────────────────────────────────────────────────

Deno.test("amount: إجمالي + number", () => {
  const results = extractAmounts("إجمالي 250.50");
  assertEquals(results.length, 1);
  assertEquals(results[0].value, 250.5);
  assertEquals(results[0].is_ambiguous, true);
});

Deno.test("amount: المطلوب: number", () => {
  const results = extractAmounts("المطلوب: 1500");
  assertEquals(results.length, 1);
  assertEquals(results[0].value, 1500);
  assertEquals(results[0].is_ambiguous, true);
});

Deno.test("amount: القيمة number", () => {
  assertEquals(extractAmounts("القيمة 300").length, 1);
});

Deno.test("amount: مبلغ number", () => {
  assertEquals(extractAmounts("مبلغ 500").length, 1);
});

Deno.test("amount: ضريبة number", () => {
  const results = extractAmounts("ضريبة 14.00");
  assertEquals(results.length, 1);
  assertEquals(results[0].value, 14);
});

Deno.test("amount: خصم number", () => {
  assertEquals(extractAmounts("خصم 50").length, 1);
});

// ── dedup ────────────────────────────────────────────────────────────────────

Deno.test("amount: does not duplicate when explicit and keyword overlap", () => {
  const results = extractAmounts("إجمالي 250.50 EGP");
  assertEquals(results.length, 1);
  assertEquals(results[0].currency, "EGP");
  assertEquals(results[0].is_ambiguous, false);
});

// ── multiple amounts ─────────────────────────────────────────────────────────

Deno.test("amount: extracts multiple amounts in one text", () => {
  const results = extractAmounts("ضريبة 14.00 EGP\nإجمالي 250");
  assertEquals(results.length, 2);
});

// ── no false positives ───────────────────────────────────────────────────────

Deno.test("amount: does not extract bare numbers without context", () => {
  assertEquals(extractAmounts("رقم الحساب 123456"), []);
});
